platform/engineering/design-docs/2020-08-17-mobile-api-design-doc.md

```diff
@@ -0,0 +1,173 @@
+
+To limit the disruption to the API it's best to perform the optimizations in a module with a `/mobile` base path. 
+This allows us to:
+
+- Add mobile authentication without changes to the core API codebase.
+- Stay out of the way as we tweak endpoints in a mobile sandbox.
+- Test out optimizations that can be later ported back to the core API.
+
+Expanding on the last point it's possible that once the mobile optimized versions of the endpoints have stabilized that we permanently
+move them back to core API. The aim is to avoid confusion about why a pereceived newer version of an endpoint would be missing fields
+without context that it was actually a 'mobile only' endpoint.
+
+#### Application Controller
+
+Authentication for the mobile app is via IAM's SSOe OAuth service. It is 100% API based and does not use callbacks or cookies. 
+The mobile `ApplicationController` would extend the core `ApplicationController` and override its authentication methods.
+
+Enabling the mobile API will expose some tech debt in the `ApplicationController` that is not a requirement to fix; but it would clean up things for all its sub-classes.
+The logic in the core `ApplicationController` that does not apply globally could move to concerns be and 
+included as needed. e.g. `render_job_id` only applies to the `DocumentsController` and the `EVSSClaimsController`. 
+
+#### Routing
+
+The mobile engine's endpoints mount at `/mobile`:
+
+```ruby
+mount Mobile::Engine, at: '/mobile'
+```
+
+Some endpoints such as facilities may not require optimization. They will pass-through via routes:
+
+```ruby
+# modules/mobile/config/routes.rb (redirect starts at the app root so this is /v0/mobile/.. pointing to )
+get '/v0/facilities/va/:id' => redirect('/v1/facilities/va/%{id}')
+```
+
+#### Request/response differences
+
+How requests and responses function within the mobile module will be isolated from the core API. However, there are improvements that could
+be ported back and it would be good to get consensus on an approach. The plan is for views, such as user profile, that currently over-fetch 
+or over-deliver, should implement fieldsets and includes in REST url params so no extra data is returned. JSON-API defines
+sparse [fieldsets](https://jsonapi.org/format/#fetching-sparse-fieldsets) and [includes](https://jsonapi.org/format/#fetching-includes) which
+can be implemented with custom code or through gems such as [JSON-API Resources](https://jsonapi-resources.com/) or [Graphiti](https://www.graphiti.dev/guides/).
+
+
+### Code Location
+The code will live in a 'mobile' directory within the [modules](https://github.com/department-of-veterans-affairs/vets-api/tree/master/modules) directory within vets-api.
+
+### Testing Plan
+To ensure changes to shared backend services will not break code, tests will be written to assert the mobile API requirements. Any breaking changes to upstream services will fail our tests.
```

Comment:
could you say more about this?

most of the `vets-api` developers will be thinking about the website primarily. how do we ensure that the output of mobile endpoints remains stable while development actively happens in the rest of the codebase? (i.e. the comment I left in your Routing section)

Reply:
We're now thinking of having our own routes and controllers and only sharing services. Request specs with [JSON schemas](http://json-schema.org/) would protect mobile responses from core API service changes.